Source/WebCore/Modules/url-pattern/URLPattern.cpp

```diff
@@ -26,29 +26,205 @@
+            String serializedBaseURL = baseURL.path().toString();
+            result.pathname = processBaseURLString(WTFMove(serializedBaseURL), type);
+        }
+
+        if (init.protocol.isEmpty()
+            && init.hostname.isEmpty()
+            && init.port.isEmpty()
+            && init.pathname.isEmpty()
+            && init.search.isEmpty()) {
+            String urlQuery = baseURL.hasQuery() ? baseURL.query().toString() : ""_s;
+            result.search = processBaseURLString(WTFMove(urlQuery), type);
+        }
+
+        if (init.protocol.isEmpty()
+            && init.hostname.isEmpty()
+            && init.port.isEmpty()
+            && init.pathname.isEmpty()
+            && init.search.isEmpty()
+            && init.hash.isEmpty()) {
+            String urlFragment = baseURL.hasFragmentIdentifier() ? baseURL.fragmentIdentifier().toString() : ""_s;
+            result.hash = processBaseURLString(WTFMove(urlFragment), type);
+        }
+    }
+
+    if (!(init.protocol.isEmpty()))
+        result.protocol = canonicalizeProtocol(WTFMove(init.protocol), type);
+
+    if (!(init.username.isEmpty()))
+        result.username = canonicalizeUsername(WTFMove(init.username), type);
+
+    if (!(init.password.isEmpty()))
+        result.password = canonicalizePassword(WTFMove(init.password), type);
+
+    if (!(init.hostname.isEmpty()))
+        result.hostname = canonicalizeHost(WTFMove(init.hostname), type);
+
+    if (!(init.port.isEmpty()))
+        result.port = canonicalizePort(WTFMove(init.port), WTFMove(init.protocol), type);
+
+    if (!(init.pathname.isEmpty())) {
+        result.pathname = init.pathname;
+
+        if (!baseURL.isNull() && baseURL.hasOpaquePath() && !isAbsolutePathname(result.pathname, type)) {
+            String baseURLPath = processBaseURLString(canonicalizePath(baseURL.path().toString(), baseURL.protocol(), type), type);
+            size_t slashIndex = baseURLPath.reverseFind("/"_s);
+
+            if (slashIndex != WTF::notFound)
+                result.pathname = makeString(baseURLPath.substring(0, slashIndex + 1), result.pathname);
+
+            result.pathname = canonicalizePath(WTFMove(result.pathname), baseURL.protocol(), type);
+        }
+    }
+
+    if (!(init.search.isEmpty()))
+        result.search = canonicalizeSearch(WTFMove(init.search), type);
+
+    if (!(init.hash.isEmpty()))
+        result.hash = canonicalizeHash(WTFMove(init.hash), type);
+
+    return result;
+}
+
 Ref<URLPattern> URLPattern::create(URLPatternInput&&, String&& baseURL, URLPatternOptions&&)
 {
     UNUSED_PARAM(baseURL);
 
     return adoptRef(*new URLPattern());
 }
 
-Ref<URLPattern> URLPattern::create(std::optional<URLPatternInput>&&, URLPatternOptions&&)
+Ref<URLPattern> URLPattern::create(std::optional<URLPatternInput>&& argInput, URLPatternOptions&&)
 {
-    return adoptRef(*new URLPattern());
+    URLPatternInit init;
+
+    if (!argInput.has_value()) {
+        // FIXME // TODO Throw Error
```

Comment:
But also, where in the spec does it say to throw? Honestly, since the spec uses an optional union with a default value, I'd expect a missing parameter to be interpreted as a default initialized first type of the union, so same as `std::holds_alternative<String>(argInput.value())` with a default-constructed String.

Reply:
It doesn't say anywhere in the spec. I was just adding it for additional safety but you're right that it would eventually be evaluated as a null string. Removed!